File: InputTestCode/Swift/SwiftTour/4_ObjectsClasses.swift
// Class definition
class Shape {
    var numberOfSides = 0
    func simpleDescription() -> String {
        return "A shape with \(numberOfSides) sides."
    }
}

// Object instantiation
var shape = Shape()

// Accessing properties
func accessClassProperties() {
    shape.numberOfSides = 7
}

// Accessing methods
func accessClassMethods() -> String {
    shape.numberOfSides = 2
    return shape.simpleDescription()
}

// Initializer and Deinitializer
// Superclasses
class NamedShape {
    var numberOfSides: Int = 0
    var name: String

    init(name: String) {
        self.name = name
    }

    deinit {
        // for cleanup before object is deallocated
    }

    func simpleDescription() -> String {
        return "A shape with \(numberOfSides) sides."
    }
}

// Subclasses
class Square: NamedShape {
    var sideLength: Double

    init(sideLength: Double, name: String) {
        self.sideLength = sideLength
        super.init(name: name)
        numberOfSides = 4
    }

    func area() -> Double {
        return sideLength * sideLength
    }

    override func simpleDescription() -> String {
        return "A square with sides of length \(sideLength)."
    }
}

// Getter/Setter
class EquilateralTriangle: NamedShape {
    var sideLength: Double = 0.0

    init(sideLength: Double, name: String) {
        self.sideLength = sideLength
        super.init(name: name)
        numberOfSides = 3
    }

    var perimeter: Double { 
        get {
            return 3.0 * sideLength
        }
        set (newValue) {
            sideLength = newValue / 3.0
        }
    }

    override func simpleDescription() -> String {
        return "An equilateral triangle with sides of length \(sideLength)."
    }
}

// Will Set
class ObservedTriangle {
    var triangle: EquilateralTriangle {
        willSet {
            print("New value for triangle!")
        }
    }
    init(size: Double, name: String) {
        triangle = EquilateralTriangle(sideLength: size, name: name)
    }
}

// Will Set accessing other variable
class TriangleAndSquare {
    var triangle: EquilateralTriangle {
        willSet {
            square.sideLength = newValue.sideLength
        }
    }
    var square: Square {
        willSet {
            triangle.sideLength = newValue.sideLength
        }
    }
    init(size: Double, name: String) {
        square = Square(sideLength: size, name: name)
        triangle = EquilateralTriangle(sideLength: size, name: name)
    }
}

// Optional accessing
let optionalSquare: Square? = Square(sideLength: 2.5, name: "optional square")
let sideLength = optionalSquare?.sideLength

File: SequalsKTranspiledCode/SwiftTour_To_Kotlin/4_ObjectsClasses.kt
package transpilation

// Class definition
class Shape {
    var numberOfSides = 0
    fun simpleDescription(): String {
        return "A shape with ${numberOfSides} sides."
    }
}

// Object instantiation
var shape = Shape()

// Accessing properties
fun accessClassProperties() {
    shape.numberOfSides = 7
}

// Accessing methods
fun accessClassMethods(): String {
    shape.numberOfSides = 2
    return shape.simpleDescription()
}

// Initializer and Deinitializer
// Superclasses
open class NamedShape(var name: String) {
    var numberOfSides: Int = 0
    {
        // for cleanup before object is deallocated
    }

    open fun simpleDescription(): String {
        return "A shape with ${numberOfSides} sides."
    }
}

// Subclasses
class Square(var sideLength: Double, name: String) : NamedShape(name) {
    init {
        numberOfSides = 4
    }

    fun area(): Double {
        return sideLength * sideLength
    }

    override fun simpleDescription(): String {
        return "A square with sides of length ${sideLength}."
    }
}

// Getter/Setter
class EquilateralTriangle(var sideLength: Double, name: String) : NamedShape(name) {
    init {
        numberOfSides = 3
    }

    var perimeter: Double
        get() {
            return 3.0 * sideLength
        }

        set(newValue) {
            sideLength = newValue / 3.0
        }

    override fun simpleDescription(): String {
        return "An equilateral triangle with sides of length ${sideLength}."
    }
}

// Will Set
class ObservedTriangle {
    var triangle: EquilateralTriangle
        set(ᏫnewValue) {
            println("New value for triangle!")
            field = ᏫnewValue
        }

    constructor(size: Double, name: String) {
        triangle = EquilateralTriangle(size, name)
    }
}

// Will Set accessing other variable
class TriangleAndSquare {
    var triangle: EquilateralTriangle
        set(ᏫnewValue) {
            square.sideLength = ᏫnewValue.sideLength
            field = ᏫnewValue
        }

    var square: Square
        set(ᏫnewValue) {
            triangle.sideLength = ᏫnewValue.sideLength
            field = ᏫnewValue
        }

    constructor(size: Double, name: String) {
        square = Square(size, name)
        triangle = EquilateralTriangle(size, name)
    }
}

// Optional accessing
val optionalSquare: Square? = Square(2.5, "optional square")
val sideLength = optionalSquare?.sideLength